### ISSUE TITLE
About the downloader

### ISSUE THREAD

**Issue by gcarreno** [LongDirtyAnimAlf/fpcupdeluxe#403]
Hey Don(@LongDirtyAnimAlf),

One thing I completely forgot to mention and it does matter a whole lot, I'm guessing, is that due to the fact that is was a quick and dirty proof of concept I did no, none, zilch, nicht, naught, zero, nada to implement ANY error reporting.

That means that if the download fails, errrmmmm.... I'm not even sure how the sate of the form is left, LOL!!!

So Don, please have a look into that and if you implement something, could I then improve my little proof of concept with your mods?

Many thanks in advance!!

Cheers,
Gus

**Comment by LongDirtyAnimAlf** [comment]
Well. Implemented with error checks. For cross-tools only.
Deep download versions needs to be both GUI as well as cmdline only.
https://github.com/LongDirtyAnimAlf/fpcupdeluxe/commit/ea377ebc91982e934e903022b84a87dc29c48e80

**Comment by gcarreno** [comment]
Hey Don(@LongDirtyAnimAlf),

> Well. Implemented with error checks. For cross-tools only.
Ok, I had to have a second look at the code cuz the first time I didn't notice the inclusion of `FSuccess`, LOL!!!

> Deep download versions needs to be both GUI as well as cmdline only.

Please elaborate on what you mean by `Deep Download` and `command line only`.
You kinda lost me here :wink: 

Cheers,
Gus

**Comment by LongDirtyAnimAlf** [comment]
The (your) normal GUI downloader has been used for the GUI frontend of fpcupdeluxe.
The deep downloader points towards the downloader of fpup itself. The command line tool. Meaning the downloader must work in a formless environment. And detect if a form is available to show its output on a dedicated form or as writen on the command line. There is also the message loop, proxy's, the sequence of events, the possibilities to interrupt the download process, the many techniques that are used to download (native, httpclient, curl, wget, ...).
Poring is very ok naturally !
I have also made some changes to the form itself (use of anchors to have good looks on systems like Darwin and exotic display managers). Please have a look.

**Comment by gcarreno** [comment]
Hey Don(@LongDirtyAnimAlf),

I've had a look at your changes and they are really nice.

I'll try and integrate it in a near future, especially the changes with the anchors. It's only been recently that I found all the possibilities or anchors and constraints and I haven't even scratched the surface. At the moment I did this example I was still in the dark regarding all the possibilities :)

I've also noticed that your latest commit limits the use to Windows and Linux.
Does that mean that my example fails in macOS?
And if so, are you planning on improving the form or the DL process to make it work under Darwin?

Cheers,
Gus
